### ISSUE TITLE
Reducing/removing the need for a temporary output edge table during simplification

### ISSUE THREAD

**Issue by molpopgen** [tskit-dev/tskit#461]
During tree sequence simplification, edges are collected and processed per parent from the input table.  New edges are stored in an "output" table, which eventually contains the full set of simplified edges.  However, the number of output edges is <= the number of input edges processed for the current parent and therefore they can be written near the start of the input table.  I've implemented this as part of molpopgen/fwdpp#242 and it can reduce the memory needs quite a bit.  It seems worthwhile to keep this in mind for tskit, too.

**Comment by jeromekelleher** [comment]
Yep, should be possible I think and probably worthwhile.

**Comment by petrelharp** [comment]
> the number of output edges is <= the number of input edges processed for the current parent 

We had thought about this, and I don't think it's always true. Do you have a proof? Consider this case:
```
parent child left right
3 0 0 0.5
3 1 0.5 1
4 2 0 1
4 3 0 1
```
The samples are `[0, 1, 2]`; simplification will remove `3` and remap `4 `to `3`.  If I'm doing it right,
we'd get
```
parent child left right
3 0 0 0.5
3 1 0.5 1
3 2 0 1
```
... so, there *are* less edges in total, but the number of edges for parent `4` (now, parent `3`), has increased. I'm not even sure the number of edges has to decrease, although I don't have a counter-example?

**Comment by molpopgen** [comment]
See the fwdpp PR for how to deal with that.

**Comment by molpopgen** [comment]
Here, specifically.  Worst case is current behavior, and typical case vastly better: https://github.com/molpopgen/fwdpp/pull/242/files#diff-9B80AA107C466C524B9329DF83299916R647

**Comment by petrelharp** [comment]
Ah, ok. I've not worked through the details in your PR, but is the idea that *usually* you're wanting to write to somewhere above where you are reading from, and you maintain a buffer in case that is not the case?

**Comment by molpopgen** [comment]
Exactly.  The <= is not a guarantee for each parent, but it is quite close
in practice.

**Comment by molpopgen** [comment]
Of course, as soon as I looked into testing overlapping generations more seriously, I have a bug in the new data structure that is for the *other* part of that PR.  :)

**Comment by benjeffery** [comment]
Closing for inactivity - reopen if this is still something you plan to do.
